# Welcome to lab_🐼s!

Last week, you did amazing work in setting up your computer to do Data Science!  Let's now do Data Science on your computer for the first time!

A few tips to remember:

- **You are not alone on your journey in learning programming!**  You have your lab Teaching Assistant, your Course Aides, your lab group, and the professors (Prof. Wade and Prof. Karle), who are all here to help you out!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

<hr style="color: #DD3403;">

# Part 0: Your Lab Group!

We find it's best when you're programming with a group of people who you can help out and can help you out!

- A key part of Data Science is being able to **communicate about your work**.  If you are the best programmer in the world, but no one uses your code, then the code you're writing is not getting run.  By learning strong communication skills, you will be able to **explain your code**, and the underlying data science concepts, to everyone -- data scientists and non-scientists alike!

- **Every lab** you will need to complete some parts of the lab **with your group**.  There will be both **"group discussions"** and **"individual reflections"** as part of every lab.

## Create Your Group

If you have not done so already, **RIGHT NOW** find *(ideally)* two other people to make a group of **two, three, or four** people.   *(We find groups of three work the best, but you can be a group of two or four -- no solo groups, and no groups of five!*)

- If you're at a group of five, split into a group of two and a group of three.  You can still help each other as a larger group, but the group questions should be answered in your smaller group of 2-4 people.

Once you have your group **introduce yourselves to each other! :)**

## Record Your Group

Since you know Python, edit the following cell to add strings about your group:

In [ ]:
# First, meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "Urvi Bhanushali"
ca1_name = "Angela"
ca2_name = "Eshaan"


# As part of your introduction, learn your lab partner's name, major, and
# learn something new about them that you never knew!
# - Groups of 3 are ideal :)
# - However groups of 2 or 4 are fine too!
partner1_name = ""
partner1_netid = ""
partner1_major = ""

partner2_name = ""
partner2_netid = ""
partner2_major = ""

partner3_name = ""
partner3_netid = ""
partner3_major = ""

<hr style="color: #DD3403;">

# Part 1: 🧑‍🤝‍🧑 Hello Dataset!

In lecture, we have already began to explore the "Party Dataset" that you created.  We will continue to use the Party Dataset all semester long, but we want even more datasets -- and we want you to come up with the questions!

In this lab, **you** will come up with questions to ask the entire class for the **"Hello Dataset"**!

## Hello Dataset Overview

Soon after this lab, your homework will be to fill out a class-wide "Hello Survey".  Right now, with your group, you will come up some of the questions that should be on this survey *(and you will do analysis on your questions next week!)*.

With your group, come up with four questions below!  Before you get started, a few notes:

- The survey will be **NOT be anonymous** and the results will be shared with the class to do analysis on, so you should **only ask questions you feel comfortable to share with the class**.
- **Questions that are inappropriate** for a classroom environment will be thrown out and you won't earn lab credit points for any question we throw out.  (Keep everything at most PG/PG-13.)
- You are going to submit **one set of questions for your whole group**.  You must decide on the questions together.

## Q1: Question with a Binary Response

A question with a **binary response** is a question that has only two answers.  For example:
- *Do you prefer cats or dogs?*
- *Were you born in Illinois?*

With your group (you must all decide on a single question together), come up with a question that you think will have **exactly 50%/50%** split between the two binary outcomes.  *(The question that gets closest to 50%/50% will get extra credit!)*

**What is your binary question?**

Do you prefer Subway or Taco Bell?

**Why do you think this is a good question?  Write at least two complete sentences on why you believe your question will have a 50%/50% split.**

This is a good question because it has a 50%/50% split towards either taco bell or subway as those two are the most debatable fast food chains in the US. People usually tend to lean towards one, so this is a binary question, more like either-or, nothing in between.

## Q2: Question with a Numeric Response

A question with a **numeric response** requires a number for an answer.  For example:

- *What is your height?*
- *What day were you born?*

With your group (you must all decide on a single question together), come up with a question and a prediction on what you think the most frequent answer will be to your question.

**What is your numeric question?**

How many credit hours are you taking this semester?

**What do you think the most frequent answer will be?**

It cannot be said for sure but for a full time student it would be anywhere between 12-18 I suppose.

**Why do you think this answer will be the most frequent answer?  Write at least one complete sentences on why you believe that it will be the most frequent answer.**

I think this would be the most frequent answer because this is what freshmans usually take if they are a full time student here at illinois.

## Q3: Question with a (non-numeric) Textual Response

A question with a (non-numeric) **textual response** requires a word or two for an answer.  For example:

- *What is your favorite sport to play?*
- *What is the location of your favorite trip/vacation?*

With your group (you must all decide on a single question together), come up with a question where at least two members of your group share the same answer.

**What is your textual response question?**

What is your favorite cuisine?

**What is the shared answer between your group?  Do you also share this answer (if not, what's your answer to the question)?**

Our answer is Italian cuisine because all of us really like the pasta and pizza!

## Q4: Open Question

One final question -- no restrictions on the format!  Is there a great question your group discussed that you did not get to use one of the first three questions but you want to know the classes' answer?

**What is your final question for the whole class?**

Milk first or Cereal first haha

## Submit Your Questions to Your TA

Your lab TAs will be grading your lab notebook for your full responses, but we want to make sure to get the survey with your questions out to everyone ASAP!

- Pick one member of your group as the "question submitter".
- If you're the question submitter, complete this Google Form for your group (requires @illinois.edu login): https://forms.gle/yMUDTbw9FSpEQ6DF7


<hr style="color: #DD3403;">

# Part 2: Importing the Course Catalog Dataset

In lecture, you learned we will use the `pandas` **library** for working with data.  Libraries are collections of functions and other objects that extend what Python can do by itself.  The `pandas` library is the most widely used library for data science.

To use the `pandas` library, we must **import** it into your notebook.  In literally every lab, you will need to import the library just like we did in lecture -- import pandas as `pd` in the following cell:

In [6]:
# Write the code to import pandas as pd:
import pandas as pd

In lecture, you saw the "Course Catalog" dataset.  Using the `read_csv` function, load the course catalog dataset.  The URL for this dataset is:
```
https://waf.cs.illinois.edu/discovery/course-catalog.csv
```

Remember, URLs passed to `read_csv` are **string** variables, so remember to use **quotation marks**. 

In [14]:
# Load the course catalog dataset into a variable named `df`:
df = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog.csv")

In [15]:
# Print out the contents of `df`:
df

Year    Term YearTerm Subject  Number                          Name  \
0      2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
1      2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
2      2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
3      2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
4      2023  Spring  2023-sp     AAS     100  Intro Asian American Studies   
...     ...     ...      ...     ...     ...                           ...   
11816  2023  Spring  2023-sp    WRIT     303          Writing Across Media   
11817  2023  Spring  2023-sp    WRIT     303          Writing Across Media   
11818  2023  Spring  2023-sp    WRIT     303          Writing Across Media   
11819  2023  Spring  2023-sp    WRIT     303          Writing Across Media   
11820  2023  Spring  2023-sp    ZULU     202            Elementary Zulu II   

                                             Description Credit Hours  \
0      Interdisciplinary introduction to the basic co...     3 hours.   
1      Interdisciplinary introduction to the basic co...     3 hours.   
2      Interdisciplinary introduction to the basic co...     3 hours.   
3      Interdisciplinary introduction to the basic co...     3 hours.   
4      Interdisciplinary introduction to the basic co...     3 hours.   
...                                                  ...          ...   
11816                    Same as INFO 303. See INFO 303.     3 hours.   
11817                    Same as INFO 303. See INFO 303.     3 hours.   
11818                    Same as INFO 303. See INFO 303.     3 hours.   
11819                    Same as INFO 303. See INFO 303.     3 hours.   
11820  Continuation of ZULU 201 with introduction of ...     5 hours.   

                                            Section Info  \
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
11816                    Same as INFO 303. See INFO 303.   
11817                    Same as INFO 303. See INFO 303.   
11818                    Same as INFO 303. See INFO 303.   
11819                    Same as INFO 303. See INFO 303.   
11820  Same as AFST 252. Participation in the languag...   

                                       Degree Attributes  ... Section Status  \
0      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
1      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
2      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
3      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
4      Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
...                                                  ...  ...            ...   
11816                       Advanced Composition course.  ...              A   
11817                       Advanced Composition course.  ...              A   
11818                       Advanced Composition course.  ...              A   
11819                       Advanced Composition course.  ...              A   
11820                                                NaN  ...              A   

       Enrollment Status                   Type Type Code Start Time  \
0                   Open  Discussion/Recitation       DIS   09:00 AM   
1                   Open  Discussion/Recitation       DIS   10:00 AM   
2                   Open  Discussion/Recitation       DIS   11:00 AM   
3                   Open  Discussion/Recitation       DIS   12:00 PM   
4                   Open  Discussion/Recitation       DIS   01:00 PM   
...                  ...                    ...       ...        ...   
11816  

## 🔬 Test Case Checkpoint 🔬

Throughout the lab, we will have "checkpoints" at key points.  Most checkpoints will include **test cases** that verify that your notebook appears to be correct.

- The test cases are **NOT** a guarantee that everything is perfect, but we try our very best to catch any common mistakes you might make!
- If you see `AssertionError` error message when running the test case, we provide a helpful hint on what may not be correct.
- If you see a party popper emoji and a message that all tests pass, you passed the checkpoint!
- **Checkpoints** are a good time to check with your group to make everyone gets to the checkpoint together!

In [16]:
## == CHECKPOINT TEST CASES for Part 1 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "Did you define a variable called df?"
assert( len(df) == 11821 ), "It looks like you did not import the correct dataset?"


## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


<hr style="color: #DD3403;">

# Part 3: STAT / CS / IS Majors

Data Science DISCOVERY is a unique course that is cross-listed across three departments: Statistics, Computer Science, and Information Sciences.  Let's explore courses from all three of those subjects!

### Puzzle 3.1

Select only the rows that have a `Subject` of `STAT`.  Assign this to the variable name `df_STAT` instead of the usual `df`.
- Make sure your result contains only statistics courses!

In [17]:
df_STAT = df[df["Subject"] == "STAT"]
df_STAT

Year    Term YearTerm Subject  Number                           Name  \
11086  2023  Spring  2023-sp    STAT     100                     Statistics   
11087  2023  Spring  2023-sp    STAT     100                     Statistics   
11088  2023  Spring  2023-sp    STAT     107         Data Science Discovery   
11089  2023  Spring  2023-sp    STAT     107         Data Science Discovery   
11090  2023  Spring  2023-sp    STAT     107         Data Science Discovery   
...     ...     ...      ...     ...     ...                            ...   
11234  2023  Spring  2023-sp    STAT     590  Individual Study and Research   
11235  2023  Spring  2023-sp    STAT     593                STAT Internship   
11236  2023  Spring  2023-sp    STAT     593                STAT Internship   
11237  2023  Spring  2023-sp    STAT     595       Preparing Future Faculty   
11238  2023  Spring  2023-sp    STAT     599                Thesis Research   

                                             Description    Credit Hours  \
11086  First course in probability and statistics at ...        3 hours.   
11087  First course in probability and statistics at ...        3 hours.   
11088  Data Science Discovery is the intersection of ...        4 hours.   
11089  Data Science Discovery is the intersection of ...        4 hours.   
11090  Data Science Discovery is the intersection of ...        4 hours.   
...                                                  ...             ...   
11234  Directed reading and research. Approved for le...   0 TO 8 hours.   
11235  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
11236  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
11237  Prepares Ph.D. students who are interested in ...        2 hours.   
11238  Approved for S/U grading only. May be repeated...  0 TO 16 hours.   

                                            Section Info  \
11086  Credit is not given for both STAT 100 and any ...   
11087  Credit is not given for both STAT 100 and any ...   
11088                         Same as CS 107 and IS 107.   
11089                         Same as CS 107 and IS 107.   
11090                         Same as CS 107 and IS 107.   
...                                                  ...   
11234  Approved for letter and S/U grading. May be re...   
11235  Approved for letter and S/U grading. Prerequis...   
11236  Approved for letter and S/U grading. Prerequis...   
11237                                                NaN   
11238  Approved for S/U grading only. May be repeated...   

                      Degree Attributes  ... Section Status  \
11086  Quantitative Reasoning I course.  ...              A   
11087  Quantitative Reasoning I course.  ...              A   
11088  Quantitative Reasoning I course.  ...              A   
11089  Quantitative Reasoning I course.  ...              A   
11090  Quantitative Reasoning I course.  ...              A   
...                                 ...  ...            ...   
11234                               NaN  ...              A   
11235                               NaN  ...              A   
11236                               NaN  ...              A   
11237                               NaN  ...              A   
11238                               NaN  ...              A   

       Enrollment Status                   Type Type Code Start Time  \
11086  Open (Restricted)                Lecture       LEC   10:00 AM   
11087             Closed                 Online       ONL   ARRANGED   
11088      CrossListOpen                Lecture       LEC   12:00 PM   
11089             Closed  Laboratory-Discussion       LBD   03:30 PM   
11090             Closed  Laboratory-Discussion       LBD   03:30 PM   
...                  ...                    ...       ...        ...   
11234             Closed      Independent Study       IND   ARRANGED   
11235  Open (Restricted)             Internship       INT   ARRANGED   
11236  Open (Restri

### Puzzle 3.2

Select only the rows that have a `Subject` of `CS`.  Assign this to the variable name `df_CS` instead of the usual `df`.
- Make sure your result contains only computer science courses!

In [18]:
df_CS = df[df["Subject"] == "CS"]
df_CS

Year    Term YearTerm Subject  Number                              Name  \
3040  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3041  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3042  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3043  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3044  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
...    ...     ...      ...     ...     ...                               ...   
3373  2023  Spring  2023-sp      CS     598                    Special Topics   
3374  2023  Spring  2023-sp      CS     598                    Special Topics   
3375  2023  Spring  2023-sp      CS     598                    Special Topics   
3376  2023  Spring  2023-sp      CS     598                    Special Topics   
3377  2023  Spring  2023-sp      CS     599                   Thesis Research   

                                            Description    Credit Hours  \
3040  Fundamental principles, concepts, and methods ...        3 hours.   
3041  Fundamental principles, concepts, and methods ...        3 hours.   
3042  Fundamental principles, concepts, and methods ...        3 hours.   
3043  Fundamental principles, concepts, and methods ...        3 hours.   
3044  Fundamental principles, concepts, and methods ...        3 hours.   
...                                                 ...             ...   
3373  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3374  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3375  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3376  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3377    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   

                                           Section Info  \
3040  Prerequisite: One of MATH 220 or MATH 221 or M...   
3041  Prerequisite: One of MATH 220 or MATH 221 or M...   
3042  Prerequisite: One of MATH 220 or MATH 221 or M...   
3043  Prerequisite: One of MATH 220 or MATH 221 or M...   
3044  Prerequisite: One of MATH 220 or MATH 221 or M...   
...                                                 ...   
3373  May be repeated in the same or separate terms ...   
3374  May be repeated in the same or separate terms ...   
3375  May be repeated in the same or separate terms ...   
3376  May be repeated in the same or separate terms ...   
3377    Approved for S/U grading only. May be repeated.   

                      Degree Attributes  ... Section Status  \
3040  Quantitative Reasoning II course.  ...              A   
3041  Quantitative Reasoning II course.  ...              A   
3042  Quantitative Reasoning II course.  ...              A   
3043  Quantitative Reasoning II course.  ...              P   
3044  Quantitative Reasoning II course.  ...              A   
...                                 ...  ...            ...   
3373                                NaN  ...              A   
3374                                NaN  ...              A   
3375                                NaN  ...              A   
3376                                NaN  ...              A   
3377                                NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3040  Open (Restricted)                Lecture       LEC   11:00 AM  11:50 AM   
3041             Closed  Laboratory-Discussion       LBD   02:00 PM  03:50 PM   
3042             Closed  Laboratory-Discussion       LBD   04:00 PM  05:50 PM   
3043             Closed  Laboratory-Discussion       LBD   06:00 PM  07:50 PM   
3044               Open  Laboratory-Discussion       LBD   09:00 AM  10:50 AM   
...                 ...                    ...       ...        ...       ...   
3373             Closed                 Online       E16   ARRANGED       NaN   
3374  Open (Restri

### Puzzle 3.3

Select only the rows that have a `Subject` of `IS`.  Assign this to the variable name `df_IS` instead of the usual `df`.
- Make sure your result contains only information science courses!

In [19]:
df_IS = df[df["Subject"] == "IS"]
df_IS

Year    Term YearTerm Subject  Number  \
6045  2023  Spring  2023-sp      IS     100   
6046  2023  Spring  2023-sp      IS     101   
6047  2023  Spring  2023-sp      IS     101   
6048  2023  Spring  2023-sp      IS     101   
6049  2023  Spring  2023-sp      IS     101   
...    ...     ...      ...     ...     ...   
6294  2023  Spring  2023-sp      IS     597   
6295  2023  Spring  2023-sp      IS     597   
6296  2023  Spring  2023-sp      IS     597   
6297  2023  Spring  2023-sp      IS     597   
6298  2023  Spring  2023-sp      IS     599   

                                                   Name  \
6045   Exploring the iSchool with a Human-Centered Lens   
6046               Introduction to Information Sciences   
6047               Introduction to Information Sciences   
6048               Introduction to Information Sciences   
6049               Introduction to Information Sciences   
...                                                 ...   
6294  Advanced Topics in Data Analytics &amp; Data S...   
6295  Advanced Topics in Data Analytics &amp; Data S...   
6296  Advanced Topics in Data Analytics &amp; Data S...   
6297  Advanced Topics in Data Analytics &amp; Data S...   
6298                                    Thesis Research   

                                            Description    Credit Hours  \
6045  This course introduces students to the School ...        1 hours.   
6046  This course provides an introduction to the fi...        3 hours.   
6047  This course provides an introduction to the fi...        3 hours.   
6048  This course provides an introduction to the fi...        3 hours.   
6049  This course provides an introduction to the fi...        3 hours.   
...                                                 ...             ...   
6294  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6295  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6296  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6297  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6298  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
6045  Prerequisite: Restricted to Majors Only; First...   
6046                                                NaN   
6047                                                NaN   
6048                                                NaN   
6049                                                NaN   
...                                                 ...   
6294  2 or 4 graduate hours. No professional credit....   
6295  2 or 4 graduate hours. No professional credit....   
6296  2 or 4 graduate hours. No professional credit....   
6297  2 or 4 graduate hours. No professional credit....   
6298  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  ... Section Status  \
6045                                 NaN  ...              A   
6046  Social & Beh Sci - Soc Sci course.  ...              A   
6047  Social & Beh Sci - Soc Sci course.  ...              A   
6048  Social & Beh Sci - Soc Sci course.  ...              A   
6049  Social & Beh Sci - Soc Sci course.  ...              A   
...                                  ...  ...            ...   
6294                                 NaN  ...              A   
6295                                 NaN  ...              A   
6296                                 NaN  ...              A   
6297                                 NaN  ...              A   
6298                                 NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
6045  Open (Restricted)     Lecture-Discussion       LCD   04:00 PM  04:50 PM   
6046             Closed  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6047               Open  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6048             Closed  Disc

### 🔬 Test Case Checkpoint 🔬


In [20]:
## == CHECKPOINT TEST CASES for Part 3 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df_CS' in vars() ), "Make certain to name the CS courses df_CS."
assert( 'df_STAT' in vars() ), "Make certain to name the STAT courses df_STAT."
assert( 'df_IS' in vars() ), "Make certain to name the IS courses df_IS."
assert( len(df_STAT[df_STAT.Subject != "STAT"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_CS[df_CS.Subject != "CS"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_IS[df_IS.Subject != "IS"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


### Puzzle 3.4

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject offers the **fewest** course sections?  The **most** course sections?


STAT offers the fewest course selections and IS offers the most course selections.

### Puzzle 3.5

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject has the highest numbered course?  Is there something similar about all of the highest numbered courses?  Write about what you observed!


All IS, CS, and STAT have 599 as the highest numbered course.

<hr style="color: #DD3403;">

# Part 4: Your Major

In this part, you will be exploring **your major** using the Course Catalog dataset.

- Use the primary subject of your **current major** (or, if you're undeclared, your intended major) for this section.
- Your answers will be different than the rest of your group!



### Puzzle 4.1:

Select only the rows that have a `Subject` of your current/intended major.  Assign this to the variable name `mydf` instead of `df`:

In [21]:
mydf = df[df["Subject"] == "CS"]
mydf

Year    Term YearTerm Subject  Number                              Name  \
3040  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3041  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3042  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3043  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3044  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
...    ...     ...      ...     ...     ...                               ...   
3373  2023  Spring  2023-sp      CS     598                    Special Topics   
3374  2023  Spring  2023-sp      CS     598                    Special Topics   
3375  2023  Spring  2023-sp      CS     598                    Special Topics   
3376  2023  Spring  2023-sp      CS     598                    Special Topics   
3377  2023  Spring  2023-sp      CS     599                   Thesis Research   

                                            Description    Credit Hours  \
3040  Fundamental principles, concepts, and methods ...        3 hours.   
3041  Fundamental principles, concepts, and methods ...        3 hours.   
3042  Fundamental principles, concepts, and methods ...        3 hours.   
3043  Fundamental principles, concepts, and methods ...        3 hours.   
3044  Fundamental principles, concepts, and methods ...        3 hours.   
...                                                 ...             ...   
3373  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3374  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3375  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3376  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3377    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   

                                           Section Info  \
3040  Prerequisite: One of MATH 220 or MATH 221 or M...   
3041  Prerequisite: One of MATH 220 or MATH 221 or M...   
3042  Prerequisite: One of MATH 220 or MATH 221 or M...   
3043  Prerequisite: One of MATH 220 or MATH 221 or M...   
3044  Prerequisite: One of MATH 220 or MATH 221 or M...   
...                                                 ...   
3373  May be repeated in the same or separate terms ...   
3374  May be repeated in the same or separate terms ...   
3375  May be repeated in the same or separate terms ...   
3376  May be repeated in the same or separate terms ...   
3377    Approved for S/U grading only. May be repeated.   

                      Degree Attributes  ... Section Status  \
3040  Quantitative Reasoning II course.  ...              A   
3041  Quantitative Reasoning II course.  ...              A   
3042  Quantitative Reasoning II course.  ...              A   
3043  Quantitative Reasoning II course.  ...              P   
3044  Quantitative Reasoning II course.  ...              A   
...                                 ...  ...            ...   
3373                                NaN  ...              A   
3374                                NaN  ...              A   
3375                                NaN  ...              A   
3376                                NaN  ...              A   
3377                                NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3040  Open (Restricted)                Lecture       LEC   11:00 AM  11:50 AM   
3041             Closed  Laboratory-Discussion       LBD   02:00 PM  03:50 PM   
3042             Closed  Laboratory-Discussion       LBD   04:00 PM  05:50 PM   
3043             Closed  Laboratory-Discussion       LBD   06:00 PM  07:50 PM   
3044               Open  Laboratory-Discussion       LBD   09:00 AM  10:50 AM   
...                 ...                    ...       ...        ...       ...   
3373             Closed                 Online       E16   ARRANGED       NaN   
3374  Open (Restri

### Puzzle 4.2:

Select **five random courses** from your major, using `mydf` and pandas functions:

In [22]:
mydf.head()

Year    Term YearTerm Subject  Number                              Name  \
3040  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3041  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3042  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3043  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3044  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   

                                            Description Credit Hours  \
3040  Fundamental principles, concepts, and methods ...     3 hours.   
3041  Fundamental principles, concepts, and methods ...     3 hours.   
3042  Fundamental principles, concepts, and methods ...     3 hours.   
3043  Fundamental principles, concepts, and methods ...     3 hours.   
3044  Fundamental principles, concepts, and methods ...     3 hours.   

                                           Section Info  \
3040  Prerequisite: One of MATH 220 or MATH 221 or M...   
3041  Prerequisite: One of MATH 220 or MATH 221 or M...   
3042  Prerequisite: One of MATH 220 or MATH 221 or M...   
3043  Prerequisite: One of MATH 220 or MATH 221 or M...   
3044  Prerequisite: One of MATH 220 or MATH 221 or M...   

                      Degree Attributes  ... Section Status  \
3040  Quantitative Reasoning II course.  ...              A   
3041  Quantitative Reasoning II course.  ...              A   
3042  Quantitative Reasoning II course.  ...              A   
3043  Quantitative Reasoning II course.  ...              P   
3044  Quantitative Reasoning II course.  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3040  Open (Restricted)                Lecture       LEC   11:00 AM  11:50 AM   
3041             Closed  Laboratory-Discussion       LBD   02:00 PM  03:50 PM   
3042             Closed  Laboratory-Discussion       LBD   04:00 PM  05:50 PM   
3043             Closed  Laboratory-Discussion       LBD   06:00 PM  07:50 PM   
3044               Open  Laboratory-Discussion       LBD   09:00 AM  10:50 AM   

     Days of Week  Room                     Building            Instructors  
3040           MW   AUD        Foellinger Auditorium  Beckman, M;Tahboub, R  
3041            M  L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  
3042            M  L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  
3043            M  L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  
3044            T  L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  

[5 rows x 27 columns]

### Puzzle 4.3:

Select **the seven lowest numbered courses** from your major:

In [23]:
mydf.tail(7)

Year    Term YearTerm Subject  Number             Name  \
3371  2023  Spring  2023-sp      CS     598   Special Topics   
3372  2023  Spring  2023-sp      CS     598   Special Topics   
3373  2023  Spring  2023-sp      CS     598   Special Topics   
3374  2023  Spring  2023-sp      CS     598   Special Topics   
3375  2023  Spring  2023-sp      CS     598   Special Topics   
3376  2023  Spring  2023-sp      CS     598   Special Topics   
3377  2023  Spring  2023-sp      CS     599  Thesis Research   

                                            Description    Credit Hours  \
3371  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3372  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3373  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3374  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3375  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3376  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3377    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   

                                           Section Info Degree Attributes  \
3371  May be repeated in the same or separate terms ...               NaN   
3372  May be repeated in the same or separate terms ...               NaN   
3373  May be repeated in the same or separate terms ...               NaN   
3374  May be repeated in the same or separate terms ...               NaN   
3375  May be repeated in the same or separate terms ...               NaN   
3376  May be repeated in the same or separate terms ...               NaN   
3377    Approved for S/U grading only. May be repeated.               NaN   

      ... Section Status  Enrollment Status                Type Type Code  \
3371  ...              A             Closed  Lecture-Discussion       S12   
3372  ...              A  Open (Restricted)  Lecture-Discussion        S6   
3373  ...              A             Closed              Online       E16   
3374  ...              A  Open (Restricted)             Lecture        L2   
3375  ...              A  Open (Restricted)  Lecture-Discussion        S8   
3376  ...              A  Open (Restricted)  Lecture-Discussion       S13   
3377  ...              A             Closed   Independent Study       IND   

     Start Time  End Time Days of Week  Room                     Building  \
3371   11:00 AM  12:15 PM           TR  1304   Siebel Center for Comp Sci   
3372   11:00 AM  12:15 PM           WF  2200     Sidney Lu Mech Engr Bldg   
3373   ARRANGED       NaN          NaN   NaN                          NaN   
3374   02:00 PM  03:15 PM           TR  1310  Digital Computer Laboratory   
3375   03:30 PM  04:45 PM           TR  2233           Everitt Laboratory   
3376   11:00 AM  12:15 PM           WF  3217           Everitt Laboratory   
3377   ARRANGED       NaN          NaN   NaN                          NaN   

                Instructors  
3371  Jabbarvand Behrouz, R  
3372            Erickson, J  
3373                 Sun, J  
3374              Forbes, M  
3375               Bates, A  
3376               Mehta, R  
3377                    NaN  

[7 rows x 27 columns]

### Puzzle 4.4:

Select **the nine highest numbered courses** from your major:

In [24]:
mydf.head(9)

Year    Term YearTerm Subject  Number                              Name  \
3040  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3041  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3042  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3043  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3044  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3045  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3046  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3047  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   
3048  2023  Spring  2023-sp      CS     101  Intro Computing: Engrg &amp; Sci   

                                            Description Credit Hours  \
3040  Fundamental principles, concepts, and methods ...     3 hours.   
3041  Fundamental principles, concepts, and methods ...     3 hours.   
3042  Fundamental principles, concepts, and methods ...     3 hours.   
3043  Fundamental principles, concepts, and methods ...     3 hours.   
3044  Fundamental principles, concepts, and methods ...     3 hours.   
3045  Fundamental principles, concepts, and methods ...     3 hours.   
3046  Fundamental principles, concepts, and methods ...     3 hours.   
3047  Fundamental principles, concepts, and methods ...     3 hours.   
3048  Fundamental principles, concepts, and methods ...     3 hours.   

                                           Section Info  \
3040  Prerequisite: One of MATH 220 or MATH 221 or M...   
3041  Prerequisite: One of MATH 220 or MATH 221 or M...   
3042  Prerequisite: One of MATH 220 or MATH 221 or M...   
3043  Prerequisite: One of MATH 220 or MATH 221 or M...   
3044  Prerequisite: One of MATH 220 or MATH 221 or M...   
3045  Prerequisite: One of MATH 220 or MATH 221 or M...   
3046  Prerequisite: One of MATH 220 or MATH 221 or M...   
3047  Prerequisite: One of MATH 220 or MATH 221 or M...   
3048  Prerequisite: One of MATH 220 or MATH 221 or M...   

                      Degree Attributes  ... Section Status  \
3040  Quantitative Reasoning II course.  ...              A   
3041  Quantitative Reasoning II course.  ...              A   
3042  Quantitative Reasoning II course.  ...              A   
3043  Quantitative Reasoning II course.  ...              P   
3044  Quantitative Reasoning II course.  ...              A   
3045  Quantitative Reasoning II course.  ...              A   
3046  Quantitative Reasoning II course.  ...              A   
3047  Quantitative Reasoning II course.  ...              A   
3048  Quantitative Reasoning II course.  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3040  Open (Restricted)                Lecture       LEC   11:00 AM  11:50 AM   
3041             Closed  Laboratory-Discussion       LBD   02:00 PM  03:50 PM   
3042             Closed  Laboratory-Discussion       LBD   04:00 PM  05:50 PM   
3043             Closed  Laboratory-Discussion       LBD   06:00 PM  07:50 PM   
3044               Open  Laboratory-Discussion       LBD   09:00 AM  10:50 AM   
3045             Closed  Laboratory-Discussion       LBD   11:00 AM  12:50 PM   
3046               Open  Laboratory-Discussion       LBD   01:00 PM  02:50 PM   
3047             Closed  Laboratory-Discussion       LBD   03:00 PM  04:50 PM   
3048               Open  Laboratory-Discussion       LBD   05:00 PM  06:50 PM   

     Days of Week  Room                     Building            Instructors  
3040           MW   AUD        Foellinger Auditorium  Beckman, M;Tahboub, R  
3041            M  L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  
3042            M  L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  
3043            M  L416  Digital Computer Laboratory  Beckman, M;Tahboub, R  
3044            T  L416  Digital Computer Laborator

### Puzzle 4.5:

Select **a random sample of 10% of the courses** from your major:

In [25]:
mydf.sample(frac = 0.1)

Year    Term YearTerm Subject  Number  \
3113  2023  Spring  2023-sp      CS     128   
3090  2023  Spring  2023-sp      CS     107   
3106  2023  Spring  2023-sp      CS     124   
3271  2023  Spring  2023-sp      CS     443   
3362  2023  Spring  2023-sp      CS     591   
3325  2023  Spring  2023-sp      CS     498   
3061  2023  Spring  2023-sp      CS     102   
3085  2023  Spring  2023-sp      CS     107   
3226  2023  Spring  2023-sp      CS     413   
3129  2023  Spring  2023-sp      CS     128   
3279  2023  Spring  2023-sp      CS     447   
3253  2023  Spring  2023-sp      CS     434   
3318  2023  Spring  2023-sp      CS     493   
3146  2023  Spring  2023-sp      CS     199   
3182  2023  Spring  2023-sp      CS     341   
3310  2023  Spring  2023-sp      CS     483   
3080  2023  Spring  2023-sp      CS     107   
3232  2023  Spring  2023-sp      CS     418   
3313  2023  Spring  2023-sp      CS     483   
3143  2023  Spring  2023-sp      CS     199   
3167  2023  Spring  2023-sp      CS     233   
3051  2023  Spring  2023-sp      CS     101   
3312  2023  Spring  2023-sp      CS     483   
3351  2023  Spring  2023-sp      CS     591   
3184  2023  Spring  2023-sp      CS     341   
3044  2023  Spring  2023-sp      CS     101   
3314  2023  Spring  2023-sp      CS     491   
3331  2023  Spring  2023-sp      CS     499   
3128  2023  Spring  2023-sp      CS     128   
3365  2023  Spring  2023-sp      CS     598   
3267  2023  Spring  2023-sp      CS     441   
3176  2023  Spring  2023-sp      CS     341   
3063  2023  Spring  2023-sp      CS     102   
3081  2023  Spring  2023-sp      CS     107   

                                                Name  \
3113             Introduction to Computer Science II   
3090                          Data Science Discovery   
3106              Introduction to Computer Science I   
3271                          Reinforcement Learning   
3362                                Advanced Seminar   
3325                                  Special Topics   
3061                        Little Bits to Big Ideas   
3085                          Data Science Discovery   
3226                          Intro to Combinatorics   
3129             Introduction to Computer Science II   
3279                     Natural Language Processing   
3253              Mobile Computing &amp; Application   
3318                          Senior Project II, ACP   
3146  Undergraduate Open Seminar in Computer Science   
3182                              System Programming   
3310                    Applied Parallel Programming   
3080                          Data Science Discovery   
3232                   Interactive Computer Graphics   
3313                    Applied Parallel Programming   
3143  Undergraduate Open Seminar in Computer Science   
3167                           Computer Architecture   
3051                Intro Computing: Engrg &amp; Sci   
3312                    Applied Parallel Programming   
3351                                Advanced Seminar   
3184                              System Programming   
3044                Intro Computing: Engrg &amp; Sci   
3314                                         Seminar   
3331                                   Senior Thesis   
3128             Introduction to Computer Science II   
3365                                  Special Topics   
3267                        Applied Machine Learning   
3176                              System Programming   
3063                        Little Bits to Big Ideas   
3081                          Data Science Discovery   

                                            Description   Credit Hours  \
3113  Continuation of CS 124. More advanced concepts...       3 hours.   
3090         Same as IS 107 and STAT 107. See STAT 107.       4 hours.   
3106  Basic concepts in computing and fundamental te...       3 hours.   
3271  Fundamental concepts and basic algorithms in R...  3 OR 4 hours.   
3362  Seminar on topics of cu

### ✔️ Checkpoint ✔️

There is no test cases for Part 4 (since everyone has different majors), but go double check your work!  There's just one section left to go! :)

<hr style="color: #DD3403;">

# Part 5: Growth of the IS major

The IS major is still a new relatively new major here at UIUC. The final puzzle is for you to analyze how much the IS major has grown since 2019.

### Puzzle 5.1: Loading the Course Catalog Dataset from 2019

Using the name `df2019` instead of `df`, load the "Fall 2019 Course Catalog Dataset" that can be found at the following URL:
```
https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv
```

In [26]:
df2019 = pd.read_csv("https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv")
df2019

Year  Term YearTerm Subject  Number  \
0     2019  Fall  2019-fa     AAS     100   
1     2019  Fall  2019-fa     AAS     105   
2     2019  Fall  2019-fa     AAS     120   
3     2019  Fall  2019-fa     AAS     199   
4     2019  Fall  2019-fa     AAS     200   
...    ...   ...      ...     ...     ...   
8584  2019  Fall  2019-fa    ZULU     202   
8585  2019  Fall  2019-fa    ZULU     403   
8586  2019  Fall  2019-fa    ZULU     404   
8587  2019  Fall  2019-fa    ZULU     405   
8588  2019  Fall  2019-fa    ZULU     406   

                                       Name  \
0              Intro Asian American Studies   
1     Introduction to Arab American Studies   
2             Intro to Asian Am Pop Culture   
3                Undergraduate Open Seminar   
4                      U.S. Race and Empire   
...                                     ...   
8584                     Elementary Zulu II   
8585                    Intermediate Zulu I   
8586                   Intermediate Zulu II   
8587                        Advanced Zulu I   
8588                       Advanced Zulu II   

                                            Description   Credit Hours  \
0     Interdisciplinary introduction to the basic co...       3 hours.   
1     Interdisciplinary introduction to the basic co...       3 hours.   
2     Introductory understanding of the way U.S. pop...       3 hours.   
3              May be repeated to a maximum of 6 hours.  1 TO 5 hours.   
4     Invites students to examine histories and narr...       3 hours.   
...                                                 ...            ...   
8584  Continuation of ZULU 201 with introduction of ...       5 hours.   
8585  Survey of more advanced grammar; emphasis on i...       4 hours.   
8586  Continuation of ZULU 403; emphasis on increasi...       4 hours.   
8587  Third year Zulu with emphasis on conversationa...       3 hours.   
8588  Continuation of Zulu 405 with increased emphas...       3 hours.   

                                           Section Info  \
0                                                   NaN   
1                                                   NaN   
2                                                   NaN   
3                                                   NaN   
4                                      Same as LLS 200.   
...                                                 ...   
8584  Same as AFST 252. Participation in the languag...   
8585                                                NaN   
8586                                                NaN   
8587                                                NaN   
8588                                                NaN   

                                      Degree Attributes  
0     Social & Beh Sci - Soc Sci, and Cultural Studi...  
1                Cultural Studies - US Minority course.  
2                Cultural Studies - US Minority course.  
3                                                   NaN  
4     Humanities - Hist & Phil, and Cultural Studies...  
...                                                 ...  
8584                                                NaN  
8585                                                NaN  
8586                                                NaN  
8587                                                NaN  
8588                                                NaN  

[8589 rows x 10 columns]

### Puzzle 5.2: Data Science Analysis

In the following cells,
- Select only the rows from the `IS` subject from the `df2019` dataset,
- Select only the rows from the `IS` subject from the `df` dataset (this contains the 2023 data),
- Complete the reflection question below (Puzzle 5.3).  Feel free to write more code to answer it.

In [30]:
df_IS_courses = df2019[df2019["Subject"] == "IS"]
df_IS_courses

Year  Term YearTerm Subject  Number  \
4843  2019  Fall  2019-fa      IS     107   
4844  2019  Fall  2019-fa      IS     199   
4845  2019  Fall  2019-fa      IS     202   
4846  2019  Fall  2019-fa      IS     310   
4847  2019  Fall  2019-fa      IS     351   
...    ...   ...      ...     ...     ...   
4930  2019  Fall  2019-fa      IS     591   
4931  2019  Fall  2019-fa      IS     592   
4932  2019  Fall  2019-fa      IS     593   
4933  2019  Fall  2019-fa      IS     594   
4934  2019  Fall  2019-fa      IS     599   

                                             Name  \
4843                       Data Science Discovery   
4844                   Undergraduate Open Seminar   
4845                     Social Aspects Info Tech   
4846                  Computing in the Humanities   
4847  The Design of Usable Information Interfaces   
...                                           ...   
4930                                    Practicum   
4931                            Independent Study   
4932                                  CAS Project   
4933                Information Sciences Practice   
4934                              Thesis Research   

                                            Description    Credit Hours  \
4843          Same as CS 107 and STAT 107. See STAT 107        4 hours.   
4844       Undergraduate Open Seminar. May be repeated.   1 TO 5 hours.   
4845       Same as INFO 202 and MACS 202. See INFO 202.        3 hours.   
4846  Explores use and application of technology to ...        3 hours.   
4847  Examines issues of Human Computer Interaction ...        3 hours.   
...                                                 ...             ...   
4930  Supervised field experience of professional-le...        2 hours.   
4931  Permits the intermediate or advanced student o...   2 TO 4 hours.   
4932  Individual study of a problem in library and i...   0 TO 8 hours.   
4933  Full-time or part-time practice of of any info...        0 hours.   
4934  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
4843          Same as CS 107 and STAT 107. See STAT 107   
4844                                                NaN   
4845       Same as INFO 202 and MACS 202. See INFO 202.   
4846                                                NaN   
4847                  Prerequisite: Sophomore standing.   
...                                                 ...   
4930  2 graduate hours. No professional credit. Appr...   
4931  2 to 4 graduate hours. No professional credit....   
4932  0 to 8 graduate hours. No professional credit....   
4933  0 graduate hours. No professional credit. Appr...   
4934  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  
4843    Quantitative Reasoning I course.  
4844                                 NaN  
4845  Social & Beh Sci - Soc Sci course.  
4846                                 NaN  
4847                                 NaN  
...                                  ...  
4930                                 NaN  
4931                                 NaN  
4932                                 NaN  
4933                                 NaN  
4934                                 NaN  

[92 rows x 10 columns]

In [31]:
df_IS = df[df["Subject"] == "IS"]
df_IS

Year    Term YearTerm Subject  Number  \
6045  2023  Spring  2023-sp      IS     100   
6046  2023  Spring  2023-sp      IS     101   
6047  2023  Spring  2023-sp      IS     101   
6048  2023  Spring  2023-sp      IS     101   
6049  2023  Spring  2023-sp      IS     101   
...    ...     ...      ...     ...     ...   
6294  2023  Spring  2023-sp      IS     597   
6295  2023  Spring  2023-sp      IS     597   
6296  2023  Spring  2023-sp      IS     597   
6297  2023  Spring  2023-sp      IS     597   
6298  2023  Spring  2023-sp      IS     599   

                                                   Name  \
6045   Exploring the iSchool with a Human-Centered Lens   
6046               Introduction to Information Sciences   
6047               Introduction to Information Sciences   
6048               Introduction to Information Sciences   
6049               Introduction to Information Sciences   
...                                                 ...   
6294  Advanced Topics in Data Analytics &amp; Data S...   
6295  Advanced Topics in Data Analytics &amp; Data S...   
6296  Advanced Topics in Data Analytics &amp; Data S...   
6297  Advanced Topics in Data Analytics &amp; Data S...   
6298                                    Thesis Research   

                                            Description    Credit Hours  \
6045  This course introduces students to the School ...        1 hours.   
6046  This course provides an introduction to the fi...        3 hours.   
6047  This course provides an introduction to the fi...        3 hours.   
6048  This course provides an introduction to the fi...        3 hours.   
6049  This course provides an introduction to the fi...        3 hours.   
...                                                 ...             ...   
6294  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6295  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6296  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6297  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6298  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
6045  Prerequisite: Restricted to Majors Only; First...   
6046                                                NaN   
6047                                                NaN   
6048                                                NaN   
6049                                                NaN   
...                                                 ...   
6294  2 or 4 graduate hours. No professional credit....   
6295  2 or 4 graduate hours. No professional credit....   
6296  2 or 4 graduate hours. No professional credit....   
6297  2 or 4 graduate hours. No professional credit....   
6298  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  ... Section Status  \
6045                                 NaN  ...              A   
6046  Social & Beh Sci - Soc Sci course.  ...              A   
6047  Social & Beh Sci - Soc Sci course.  ...              A   
6048  Social & Beh Sci - Soc Sci course.  ...              A   
6049  Social & Beh Sci - Soc Sci course.  ...              A   
...                                  ...  ...            ...   
6294                                 NaN  ...              A   
6295                                 NaN  ...              A   
6296                                 NaN  ...              A   
6297                                 NaN  ...              A   
6298                                 NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
6045  Open (Restricted)     Lecture-Discussion       LCD   04:00 PM  04:50 PM   
6046             Closed  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6047               Open  Discussion/Recitation       DIS   11:00 AM  12:20 PM   
6048             Closed  Disc

In [29]:
# Feel free to add more code cells if needed by hovering over the area below the cell and pressing "+ Code"!

### Puzzle 5.3: Reflection on the Growth of Information Science

In the follow cell, write **at least one complete sentence explaining what you found** related to the growth of the IS program.  In your answer, **use actual data** (ex: by how much did it grow?)

The IS program actually has a huge growth since 2019 because it added so many other courses from 92 to 254.

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)